src/sdmmc.rs

```diff
@@ -534,6 +535,17 @@ where
     }
 }
 
+impl<SPI, CS> Drop for BlockSpi<'_, SPI, CS>
+where
+    SPI: embedded_hal::blocking::spi::Transfer<u8>,
+    <SPI as embedded_hal::blocking::spi::Transfer<u8>>::Error: core::fmt::Debug,
+    CS: embedded_hal::digital::v2::OutputPin,
+{
+    fn drop(&mut self) {
+        self.deinit()
```

Comment:
At the moment this is fine since `deinit` cannot fail.
However, its docs already hint at flushing data if needed during `deinit`, which can fail.
The problem is that `Drop` cannot return an error so if we would add some data flushing during `deinit`, this `Drop` implementation would need to `panic` on error.
Have you thought about this?

Reply:
That's a good observation, and I have one example of the same problem in the standard library:

https://doc.rust-lang.org/src/std/io/buffered/bufwriter.rs.html#668

Here, the error is dropped. It's not an amazing solution though, so I have another idea to the same effect: expose the BlockSpi inside a closure instead of in the "guard" pattern. Like this:

```
let result = sdmmcspi.acquire(|block: BlockSpi| { do_my_stuff() });
```

In this case, the result from `acquire` would contain the result of the final flush, e.g. `Result<ClosureRet, FlushFailed>`.

Closures can be nasty to use, so I see this more as an addition to the guard pattern than a replacement.

Reply:
I have implemented that in 6005533f424fa4cb48bd89f713d8acb4d39c1c2e but it's not actually necessary as long as the SPI device deinit is infallible, so I didn't add that commit to this proposal.

Reply:
I see. It would need to be documented that the `with_block` method flushes, though.
However, the problem would still be there, although mitigated, if `deinit` did anything fallible, independently of whether the operations and their flush ran within a closure.
I cannot think of a solution to this, though. We would need https://github.com/rust-lang/rfcs/issues/814

I still think this approach is more ergonomic than the status-quo.
Could you change the documentation of `deinit` to say that any flush must be done beforehand and `deinit` must always be kept infallible due to this drop?

Reply:
Updated with more docs. I think now it's a bit split-brain, because I think `deinit` should be called manually before drops, but it *also* should attempt to flush, following the logic that a failed flush attempted always is better then not enforcing any flush at all.